### PR TITLE
8286441: Remove mode parameter from jdk.internal.perf.Perf.attach()

### PR DESCRIPTION
**Pull request by iklam**
The `mode` parameter for ` jdk.internal.perf.Perf.attach()` has never supported the value `"rw"` since the source code was imported to the openjdk repo more than 15 years ago. In fact HotSpot throws `IllegalArgumentException` when such a mode is specified.

It's unlikely such a mode will be required for future enhancements. Support for `"rw"` is removed. The "mode" parameter is also removed, since now `"r"` is the only supported mode.

I also cleaned up related code in the JDK and HotSpot.

Testing:
- Passed tiers 1 ~ 5

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issue
 * [JDK-8286441](https://bugs.openjdk.java.net/browse/JDK-8286441): Remove mode parameter from jdk.internal.perf.Perf.attach()

### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**) ⚠️ Review applies to [22c22c30](https://git.openjdk.java.net/jdk/pull/8622/files/22c22c3039e991c3793887b03826f8668c40b824)
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8622/head:pull/8622` \
`$ git checkout pull/8622`

Update a local copy of the PR: \
`$ git checkout pull/8622` \
`$ git pull https://git.openjdk.java.net/jdk pull/8622/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8622`

View PR using the GUI difftool: \
`$ git pr show -t 8622`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8622.diff">https://git.openjdk.java.net/jdk/pull/8622.diff</a>

</details>

### CODE REVIEW

File: src/jdk.internal.jvmstat/share/classes/sun/jvmstat/perfdata/monitor/protocol/file/PerfDataBuffer.java [hunk 1]
```diff
@@ -53,22 +53,11 @@ public class PerfDataBuffer extends AbstractPerfDataBuffer {
...
-                bb = fc.map(FileChannel.MapMode.READ_ONLY, 0L, (int)fc.size());
-            } else if (mode.equals("rw")) {
-                bb = fc.map(FileChannel.MapMode.READ_WRITE, 0L, (int)fc.size());
-            } else {
-                throw new IllegalArgumentException("Invalid mode: " + mode);
-            }
-
+            FileChannel fc = new RandomAccessFile(f, "r").getChannel();
+            ByteBuffer bb = fc.map(FileChannel.MapMode.READ_ONLY, 0L, (int)fc.size());
             fc.close();               // doesn't need to remain open
```
Comment (AlanBateman) [inline src/jdk.internal.jvmstat/share/classes/sun/jvmstat/perfdata/monitor/protocol/file/PerfDataBuffer.java:60]: I think you can change this to:

```
        try (FileChannel fc = FileChannel.open(f.toPath())) {
            ByteBuffer bb = ...
            createPerfDataBuffer(bb, 0);
        }
```

**Comment by AlanBateman** [comment]
I skimmed through the changes and I think they look okay. In the distant past there were tools outside of the JDK that used the jvmstat API directly. It's possible that VisualVM still does but it would only compile/run if --add-exports is used to export the sun.jvmstat.* packages. So it might be that dropping the parameter from a method in RemoteHost is noticed and I think that is okay because this package is not exported and is not meant to be used by code outside of the JDK.

File: src/jdk.jstatd/share/classes/sun/tools/jstatd/RemoteHostImpl.java [hunk 2]
```diff
@@ -67,17 +67,13 @@ public RemoteHostImpl(int rmiPort) throws MonitorException {
         monitoredHost.addHostListener(this);
     }
 
-    public RemoteVm attachVm(int lvmid, String mode)
+    public RemoteVm attachVm(int lvmid)
                     throws RemoteException, MonitorException {
         Integer v = lvmid;
         RemoteVm stub = null;
         StringBuilder sb = new StringBuilder();
```
Comment (cl4es) [inline src/jdk.jstatd/share/classes/sun/tools/jstatd/RemoteHostImpl.java:74]:
```suggestion
        String vmidStr = "local://" + lvmid + "@localhost";
```

File: src/hotspot/share/runtime/perfMemory.hpp [hunk 3]
```diff
@@ -148,7 +143,7 @@ class PerfMemory : AllStatic {
 
     // methods for attaching to and detaching from the PerfData
     // memory segment of another JVM process on the same system.
-    static void attach(const char* user, int vmid, PerfMemoryMode mode,
+    static void attach(const char* user, int vmid,
```
Comment (cl4es) [inline src/hotspot/share/runtime/perfMemory.hpp:146]: One line?

File: src/hotspot/share/prims/perf.cpp [hunk 4]
```diff
@@ -80,13 +80,8 @@ PERF_ENTRY(jobject, Perf_Attach(JNIEnv *env, jobject unused, jstring user, int v
...
   }
 
-  if (mode != PerfMemory::PERF_MODE_RO &&
-      mode != PerfMemory::PERF_MODE_RW) {
-    THROW_0(vmSymbols::java_lang_IllegalArgumentException());
-  }
-
   // attach to the PerfData memory region for the specified VM
-  PerfMemory::attach(user_utf, vmid, (PerfMemory::PerfMemoryMode) mode,
+  PerfMemory::attach(user_utf, vmid,
```
Comment (cl4es) [inline src/hotspot/share/prims/perf.cpp:84]: One line?

File: src/hotspot/os/windows/perfMemory_windows.cpp [hunk 5]
```diff
@@ -1796,7 +1778,7 @@ void PerfMemory::delete_memory_region() {
 // the indicated process's PerfData memory region into this JVMs
 // address space.
 //
-void PerfMemory::attach(const char* user, int vmid, PerfMemoryMode mode,
+void PerfMemory::attach(const char* user, int vmid,
```
Comment (cl4es) [inline src/hotspot/os/windows/perfMemory_windows.cpp:1781]: One line?

**Review by cl4es — APPROVED** [review]
Nice cleanup!

(Some stylistic suggestions inline, feel free to ignore)

File: src/jdk.internal.jvmstat/share/classes/sun/jvmstat/perfdata/monitor/protocol/file/PerfDataBuffer.java (hunk 1 above)
Comment (iklam) [reply src/jdk.internal.jvmstat/share/classes/sun/jvmstat/perfdata/monitor/protocol/file/PerfDataBuffer.java]: Fixed.

File: src/hotspot/os/windows/perfMemory_windows.cpp (hunk 5 above)
Comment (iklam) [reply src/hotspot/os/windows/perfMemory_windows.cpp:1780]: Fixed

File: src/hotspot/share/prims/perf.cpp (hunk 4 above)
Comment (iklam) [reply src/hotspot/share/prims/perf.cpp]: Fixed

File: src/hotspot/share/runtime/perfMemory.hpp (hunk 3 above)
Comment (iklam) [reply src/hotspot/share/runtime/perfMemory.hpp]: Fixed

File: src/jdk.jstatd/share/classes/sun/tools/jstatd/RemoteHostImpl.java (hunk 2 above)
Comment (iklam) [reply src/jdk.jstatd/share/classes/sun/tools/jstatd/RemoteHostImpl.java:73]: Fixed

**Comment by iklam** [comment]
> I skimmed through the changes and I think they look okay. In the distant past there were tools outside of the JDK that used the jvmstat API directly. It's possible that VisualVM still does but it would only compile/run if --add-exports is used to export the sun.jvmstat.* packages. So it might be that dropping the parameter from a method in RemoteHost is noticed and I think that is okay because this package is not exported and is not meant to be used by code outside of the JDK.

The APIs changed by this PR are:

- sun.jvmstat.monitor.remote.RemoteHost::attachVm
- sun.jvmstat.monitor.VmIdentifier::getMode
- sun.jvmstat.monitor.HostIdentifier::getMode
- jdk.internal.perf.Perf::attach

I checked the latest VisualVM source code. Some of the above classes are referenced, but none of the affected APIs are called.

**Review by AlanBateman — APPROVED** [review]

**Comment by iklam** [comment]
Thanks to @AlanBateman and @cl4es for the review.
/integrate

**Pull request closed without merging**